File: packages/inderun-web/src/provider.ts
import type { TaskRequest, TaskResult } from "@independo/inderun-contracts";
import type { HostServices } from "./host.js";

/**
 * Static metadata descriptor defining the capabilities, requirements, and constraints
 * of a provider. This is used by the PolicyEngine and Router to select candidate providers.
 */
export interface ProviderDescriptor {
  /**
   * Unique identifier of the provider (e.g., 'openai', 'apple-foundation-models').
   */
  id: string;
  /**
   * Execution target type: 'local' (on-device), 'edge', or 'cloud'.
   */
  type: "local" | "edge" | "cloud";
  /**
   * Connection transport layer protocol.
   */
  transport: "in_process" | "system_service" | "http" | "sse" | "realtime";
  /**
   * Optional streaming style, if token or chunk streaming is supported.
   */
  streamingStyle?: "tokens" | "chunks" | "snapshots";

  /**
   * Flags representing the features and modes supported by this provider.
   */
  supports: {
    /** Whether the provider supports Mode 1 run (request/response). */
    run: boolean;
    /** Whether the provider supports Mode 2 stream. */
    streaming: boolean;
    /** Whether the provider supports Mode 3 realtime session. */
    realtime: boolean;
    /** Whether the provider supports tool calling definition. */
    tools: boolean;
    /** Whether the provider outputs reasoning events alongside contents. */
    reasoningEvents: boolean;
    /** Whether the provider supports structured outputs (JSON schema constraints). */
    structuredOutput: boolean;
    /** Whether the provider supports multimodal inputs (images, audio, etc.). */
    multimodal: boolean;
  };

  /**
   * Cancellation capabilities support:
   * - 'hard': fully stops processing and interrupts remote operations.
   * - 'soft': stops reading the stream locally, but the remote execution continues.
   * - 'none': cancellation is unsupported.
   */
  cancel: "hard" | "soft" | "none";

  /**
   * List of supported task kinds (e.g., 'text_to_text', 'embeddings').
   */
  tasks: string[];

  /**
   * Optional input/output resource boundaries.
   */
  limits?: {
    maxInputTokens?: number;
    maxOutputTokens?: number;
    maxImageBytes?: number;
    maxAudioSeconds?: number;
  };

  /**
   * Privacy settings regarding data egress.
   */
  privacy?: {
    /** Set to true if the model input leaves the host device. */
    dataLeavesDevice: boolean;
    /** Geographic regions where the cloud service executes. */
    regions?: string[];
  };
}

/**
 * Dynamic capability snapshot evaluated at runtime.
 */
export interface ProviderDynamicCapabilities {
  /**
   * Whether the provider is available and functional on the host device now.
   */
  available: boolean;
  /**
   * Detail message if availability check fails.
   */
  reason?: string;
}

/**
 * Execution context passed to provider run commands.
 */
export interface RunContext {
  /**
   * Unique execution run ID.
   */
  runId: string;
  /**
   * Host service boundary for provider adapters. Cloud adapters use this to
   * resolve authContextRef through secure storage and dispatch HTTP requests.
   */
  hostServices: HostServices;
}

/**
 * Pluggable execution adapter contract that wraps a specific model runtime
 * (system frameworks, local runtimes, or cloud APIs) and exposes normalized APIs.
 */
export interface ProviderAdapter {
  /**
   * Returns static provider descriptor metadata.
   */
  describe(): ProviderDescriptor;
  /**
   * Performs dynamic capability check to determine if the provider can execute now.
   */
  capabilities(host: HostServices): Promise<ProviderDynamicCapabilities>;
  /**
   * Executes a task request in Mode 1 (request/response).
   */
  run(req: TaskRequest, ctx: RunContext): Promise<TaskResult>;
}

File: ios/IndeRun/Sources/IndeRunCore/Provider.swift
import Foundation
import IndeRunContracts

// MARK: - ProviderDescriptor
public struct ProviderDescriptor: Codable, Sendable {
    public enum ProviderType: String, Codable, Sendable {
        case local
        case edge
        case cloud
    }
    
    public enum TransportType: String, Codable, Sendable {
        case inProcess = "in_process"
        case systemService = "system_service"
        case http
        case sse
        case realtime
    }
    
    public enum StreamingStyle: String, Codable, Sendable {
        case tokens
        case chunks
        case snapshots
    }
    
    public struct SupportsCapabilities: Codable, Sendable {
        public let run: Bool
        public let streaming: Bool
        public let realtime: Bool
        public let tools: Bool
        public let reasoningEvents: Bool
        public let structuredOutput: Bool
        public let multimodal: Bool
        
        public init(
            run: Bool,
            streaming: Bool,
            realtime: Bool,
            tools: Bool,
            reasoningEvents: Bool,
            structuredOutput: Bool,
            multimodal: Bool
        ) {
            self.run = run
            self.streaming = streaming
            self.realtime = realtime
            self.tools = tools
            self.reasoningEvents = reasoningEvents
            self.structuredOutput = structuredOutput
            self.multimodal = multimodal
        }
    }
    
    public enum CancelSemantics: String, Codable, Sendable {
        case hard
        case soft
        case none
    }
    
    public struct ResourceLimits: Codable, Sendable {
        public let maxInputTokens: Int?
        public let maxOutputTokens: Int?
        public let maxImageBytes: Int?
        public let maxAudioSeconds: Int?
        
        public init(
            maxInputTokens: Int? = nil,
            maxOutputTokens: Int? = nil,
            maxImageBytes: Int? = nil,
            maxAudioSeconds: Int? = nil
        ) {
            self.maxInputTokens = maxInputTokens
            self.maxOutputTokens = maxOutputTokens
            self.maxImageBytes = maxImageBytes
            self.maxAudioSeconds = maxAudioSeconds
        }
    }
    
    public struct PrivacyDescriptor: Codable, Sendable {
        public let dataLeavesDevice: Bool
        public let regions: [String]?
        
        public init(dataLeavesDevice: Bool, regions: [String]? = nil) {
            self.dataLeavesDevice = dataLeavesDevice
            self.regions = regions
        }
    }
    
    public let id: String
    public let type: ProviderType
    public let transport: TransportType
    public let streamingStyle: StreamingStyle?
    public let supports: SupportsCapabilities
    public let cancel: CancelSemantics
    public let tasks: [String]
    public let limits: ResourceLimits?
    public let privacy: PrivacyDescriptor?
    
    public init(
        id: String,
        type: ProviderType,
        transport: TransportType,
        streamingStyle: StreamingStyle? = nil,
        supports: SupportsCapabilities,
        cancel: CancelSemantics,
        tasks: [String],
        limits: ResourceLimits? = nil,
        privacy: PrivacyDescriptor? = nil
    ) {
        self.id = id
        self.type = type
        self.transport = transport
        self.streamingStyle = streamingStyle
        self.supports = supports
        self.cancel = cancel
        self.tasks = tasks
        self.limits = limits
        self.privacy = privacy
    }
}

// MARK: - ProviderDynamicCapabilities
public struct ProviderDynamicCapabilities: Codable, Sendable {
    public let available: Bool
    
    public init(available: Bool) {
        self.available = available
    }
}

// MARK: - RunContext
public struct RunContext: Sendable {
    public let runId: String
    public let hostServices: HostServices
    
    public init(runId: String, hostServices: HostServices) {
        self.runId = runId
        self.hostServices = hostServices
    }
}

// MARK: - ProviderAdapter Protocol
public protocol ProviderAdapter: Sendable {
    func describe() -> ProviderDescriptor
    func capabilities(host: HostServices) async -> ProviderDynamicCapabilities
    func run(request: TaskRequest, context: RunContext) async throws -> TaskResult
}
